src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+                corresponding to the determined edge positions.
+            shape ({edges, trains}, optional): Whether to
+                keep a single edge dimension (`edges`, default) or group
+                edges by train (`trains`). The latter option return
+                labelled xarray objects, while the former returns pandas
+                series or dataframe.
+            squeeze_edges (bool, optional): Whether to minimize the edge
+                axis lengh for xarray objects, True by default and
+                ignored when pandas objects are returned.
+            pasha_ctx (pasha.MapContext, optional): Pasha context to run
+                parallelized operations, pasha.ProcessContext by default.
+
+            Any further keyword arguments are passed to the edge finding
+            method.
+
+        Returns:
+            result (pd.Series): Edge positions, only if edge shaped and
+                without amplitudes.
+
+            reuslt (pd.DataFrame): Edge positions and pulse heights,
+                only if edge shaped and with amplitudes.
+
+            result (xarray.DataArray) Edge positions, only if train
+                shaped and without amplitudes.
+
+            result (xarray.Dataset): Edge positions and pulse heights,
+                only if train shaped and with amplitudes.
+        """
+
+        edge_func = self._validate_edge_method(edge_func, edge_kw)
+
+        if max_edges is None:
+            max_edges = self._raw_key.entry_shape[0] // 5000
+
+        # Set-up pasha for parallel processing.
+        psh = self._validate_pasha(pasha_ctx)
+        edges = psh.alloc(shape=(self._raw_key.shape[0], max_edges),
+                          dtype=self._preprocess_kw['dtype'], fill=np.nan)
+
+        if with_amplitudes:
+            amplitudes = psh.alloc(like=edges, fill=np.nan)
+        else:
+            amplitudes = np.zeros(max_edges, dtype=edges.dtype)
+
+        def digitize_edges(wid, index, train_id, trace):
+            edge_func(signal=self.preprocess(trace, **self._preprocess_kw),
+                      edges=edges[index],
+                      amplitudes=(amplitudes[index] if with_amplitudes
+                                  else amplitudes),
+                      **edge_kw)
+
+        psh.map(digitize_edges, self._raw_key)
+
+        if sample_dim == 'time':
+            edges *= self.sampling_period
+
+        if shape == 'edges':
+            index_or_coords = pd.Index(self._raw_key.train_id_coordinates(),
+                                       name='trainId')
+
+        elif shape == 'trains':
+            index_or_coords = {'trainId': self._raw_key.train_id_coordinates()}
+
+        else:
+            raise ValueError('shape must be one of `edges`, `trains`')
+
+        return self._shape_edge_result(edges, amplitudes, index_or_coords,
+                                       shape, squeeze_edges, with_amplitudes)
+
+    def pulse_edges(self, sample_dim='sample',
```

Comment:
This signature is already the result of as much pruning as seemed sensible, maybe someone has more ideas.